src/states/node.rs

```diff
@@ -2541,15 +2370,22 @@ impl Node {
             let our_merged_section: BTreeSet<_> = our_sections
                 .iter()
                 .chain(their_sections.iter())
-                .filter(|&(prefix, _)| prefix.popped() == self.our_prefix().popped())
+                .filter(|&(ver_pfx, _)| ver_pfx.prefix().is_extension_of(&merge_prefix))
                 .flat_map(|(_, peers)| peers)
                 .map(|peer| *peer.name())
                 .collect();
-            self.process_own_section_merge(our_prefix, our_sections, &our_merged_section, outbox)?;
+            let version = our_sections
```

Comment:
To avoid doing more or less the same iteration/filter twice, I'd have written this as:

```rust
let mut version = 1;
let mut our_merged_section = BTreeSet::new();
for (ver_pfx, peers) in
    our_sections
        .iter()
        .chain(their_sections.iter())
        .filter(|&(ver_pfx, _)| ver_pfx.prefix().extends(&merge_prefix)) {
    version = cmp::max(version, ver_pfx.version() + 1);
    our_merged_section.append(&mut peers
                                       .into_iter()
                                       .map(|peer| *peer.name())
                                       .collect());
}
```
But that's probably not any clearer, so feel free to ignore :)